features/123-chat-push-notification/api.md

```diff
@@ -1,158 +1,108 @@
 
-```javascript
-export function afterMessageSent(callback) {
-  registry.registerHandler('chat:after_message_sent', (req) => {
-    var conversation = blahblahblah ;
-    var message = blahblahblah;
-    callback(message, conversation);
-  }, {});
-}
-```
+### Main Payload
 
-### Changes on Plugin (Hook Wrapper)
-- Wrapper of op wrapper
+Main Payload is a collection of payloads.
 
-- Example:
+|key | type |description           |default|
+|----|------|----------------------|-------|
+|platform|dict|See platform payload|null|
+|message|dict|See message payload|null|
+|data|dict|Custom dictionary by developer|null|
+|custom|dict|See custom payload|null|
+|dry\_run|boolean|dry run only if true|false|
 
-```python
-def after_message_sent(*args, **kwargs):
-    def chat_op(func):
-        _registry.register_op('chat:after_message_sent', func, *args, **kwargs)
-        return func
-    return chat_op
-```
 
 
-### Changes on Plugin (Database Operation)
-- Send message to call specific hook after database operation
-- Example:
+### Platform Payload
 
-``` python
-container.send_action('chat:after_conversation_hook', {'roles': message, 'conversation': conversation} )
-```
+|key | type |description|default|
+|----|------|-----------|-------|
+|android|boolean|send to android device if true|true|
+|ios|boolean|send to android device if true|true|
+
+### Message Payload
+
+|key|type|description|default|
+|---|----|-----------|-------|
+|body|string|message body|null|
+|title|string|message title|null|
+|subtitle|string|message subtitle|null|
+|sound|string|message sound|null|
+|title\_loc\_key|string|key to localized string|null|
+|title\_loc\_args|list of string|arguments of string replacement of title\_loc\_key|null|
+|priority|string|priority of message|"normal"|
+|expiry|int|time to live in seconds| 2419200 (4 weeks)|
 
-### Changes on API at skygear-server
 
-Nil
+### Custom Payload
+|key|type|description|default|
+|---|----|-----------|-------|
+|android\_channel\_id|string| |null|
+|android\_click\_action|string| |null|
+|android\_body\_loc\_key\_action|string| |null|
+|android\_body\_loc\_key\_args|string| |null|
+|android\_led\_color|string| |null|
+|android\_icon|string| |null|
+|ios\_badge|integer| |null|
+|ios\_content\_available|boolean| |null|
+|ios\_mutable\_content|boolean| |null|
+|ios\_launch\_image|string| |null|
+|ios\_category|string| |null|
 
-### Database Schema
 
-Nil
 
-### Migration
+## Implementation Details
+
+
+
+### Changes on SDK
+1. SDK needs to provide with simple interface for minimal push notification.
+
+
+Sample Interface:
+
+```javascript
+function sendToUser(users, payload)
+```
+
+Sample Call:
+```javascript
+
+function sendToUser(users, {
+  'message': {
+    'title': 'Hello World',
+    'body': 'Hello from oursky.'
+  }
+});
+```
+
+2. SDK supports `users` in either list of user ids and list of user objects.
+3. SDK (in Android and iOS) should have payload builder class for auto complete.
```

Comment:
Actaully when we design the new push notification, we will want to include web platform. i.e. browser push notification.

Reply:
I asked @carmenlau that we wanted to target at iOS and Android at this moment, am I right? And if we want to also support web push notification.

1. Do I need to scope out parameters for browsers?
2. Is it really supported by Skygear server?

Reply:
Okay, agree we can focus on two main target platform for now.